File: scala/ch01/ex1.31.scala
def prod(term: (Int) => Int, a: Int, next: (Int) => (Int), b: Int) = {
  def p(a2: Int, accum: Int): Int = if (a2 > b) accum else
      p(next(a2), term(a2) * accum)
  p(a, 1)
}

def prod2(term: (Int) => Int, a: Int, next: (Int) => (Int), b: Int): Int = 
  if (a > b) 1 else 
    (term(a) * prod2(term, next(a), next, b))

def identity[A](n: A) = n
def inc(n: Int) = n + 1

def fact(n: Int) = prod(identity, 1, inc, n)
def fact2(n: Int) = prod2(identity, 1, inc, n)

import org.scalatest._
import org.scalatest.matchers._

val factVals = Array(1, 2, 6, 24, 120, 720, 5040)
  
object factorialSpec extends Spec with ShouldMatchers {
  describe ("Factorials computed with recursive 'prod'") {
    it ("should calculate the correct values") {
      (1 to 7) foreach { n => fact(n) should equal (factVals(n-1)) }
    }
  }
  describe ("Factorials computed with iterative 'prod'") {
    it ("should calculate the correct values") {
      (1 to 7) foreach { n => fact2(n) should equal (factVals(n-1)) }
    }
  }
}

factorialSpec execute

File: clojure/ch01/ex1.31.clj
(ns sicp.ch01 (:use clojure.test))

(defn prod [term a next b]
  (if (> a b)
    1.0
    (* (term a) (prod term (next a) next b))))

; tail recursive
(defn prod2 [term a next b]
  (loop [a2 a accum 1.0]
    (if (> a2 b)
      accum
      (recur (next a2) (* accum (term a2))))))

(defn ident [n] n)
(defn increment [n] (+ n 1))

(defn fact  [n] (prod  ident 1 increment n))
(defn fact2 [n] (prod2 ident 1 increment n))

(deftest test-fact
  (is (= 1    (fact 1)))
  (is (= 2    (fact 2)))
  (is (= 6    (fact 3)))
  (is (= 24   (fact 4)))  
  (is (= 120  (fact 5)))  
  (is (= 720  (fact 6)))
  (is (= 5040 (fact 7))))

(deftest test-fact2
  (is (= 1    (fact2 1)))
  (is (= 2    (fact2 2)))
  (is (= 6    (fact2 3)))
  (is (= 24   (fact2 4)))  
  (is (= 120  (fact2 5)))  
  (is (= 720  (fact2 6)))
  (is (= 5040 (fact2 7))))

(run-tests)

; fact2 is actually slower! That's believable for low "n", because of the 
; overhead of trampoling, but surprising for large "n".
(time (fact  7))
(time (fact2 7))
(time (fact  200))
(time (fact2 200))

